pkg/upgrade/job/job.go

```diff
@@ -138,116 +135,46 @@ func NewUpgradeJob(plan *upgradeapiv1.Plan, nodeName, controllerName string) *ba
-						Args:    plan.Spec.Upgrade.Args,
-						VolumeMounts: []corev1.VolumeMount{
-							{Name: "host-root", MountPath: "/host"},
-							{Name: "pod-info", MountPath: "/run/system-upgrade/pod", ReadOnly: true},
-						},
-						Env: []corev1.EnvVar{{
-							Name:  "SYSTEM_UPGRADE_PLAN_NAME",
-							Value: plan.Name,
-						}, {
-							Name:  "SYSTEM_UPGRADE_PLAN_LATEST_HASH",
-							Value: plan.Status.LatestHash,
-						}, {
-							Name:  "SYSTEM_UPGRADE_PLAN_LATEST_VERSION",
-							Value: plan.Status.LatestVersion,
-						}, {
-							Name: "SYSTEM_UPGRADE_NODE_NAME",
-							ValueFrom: &corev1.EnvVarSource{
-								FieldRef: &corev1.ObjectFieldSelector{
-									FieldPath: "spec.nodeName",
-								},
-							},
-						}, {
-							Name: "SYSTEM_UPGRADE_POD_NAME",
-							ValueFrom: &corev1.EnvVarSource{
-								FieldRef: &corev1.ObjectFieldSelector{
-									FieldPath: "metadata.name",
-								},
-							},
-						}, {
-							Name: "SYSTEM_UPGRADE_POD_UID",
-							ValueFrom: &corev1.EnvVarSource{
-								FieldRef: &corev1.ObjectFieldSelector{
-									FieldPath: "metadata.uid",
 								},
-							},
-						}},
-					}},
+							}),
+							withSecrets(plan.Spec.Secrets),
+							withPlanEnvironment(plan.Name, plan.Status),
+						),
+					},
 				},
 			},
 		},
 	}
 
-	for _, secret := range plan.Spec.Secrets {
-		secretVolumeName := name.SafeConcatName("secret", secret.Name)
-		secretVolumePath := secret.Path
-		if secretVolumePath == "" {
-			secretVolumePath = filepath.Join("/run/system-upgrade/secrets", secret.Name)
-		} else if secretVolumePath[0:1] != "/" {
-			secretVolumePath = filepath.Join("/run/system-upgrade/secrets", secretVolumePath)
-		}
-		job.Spec.Template.Spec.Volumes = append(job.Spec.Template.Spec.Volumes, corev1.Volume{
-			Name: secretVolumeName,
-			VolumeSource: corev1.VolumeSource{
-				Secret: &corev1.SecretVolumeSource{
-					SecretName: secret.Name,
-				},
-			},
-		})
-		job.Spec.Template.Spec.Containers[0].VolumeMounts = append(job.Spec.Template.Spec.Containers[0].VolumeMounts, corev1.VolumeMount{
-			Name:      secretVolumeName,
-			MountPath: secretVolumePath,
-			ReadOnly:  true,
-		})
+	// first, we prepare
+	if plan.Spec.Prepare != nil {
+		job.Spec.Template.Spec.InitContainers = append(job.Spec.Template.Spec.InitContainers,
+			container("prepare", *plan.Spec.Prepare,
+				withSecrets(plan.Spec.Secrets),
+				withPlanEnvironment(plan.Name, plan.Status),
+			),
+		)
 	}
 
+	// then we cordon/drain
 	cordon, drain := plan.Spec.Cordon, plan.Spec.Drain
 	if drain != nil {
 		args := []string{"drain", nodeName, "--pod-selector", `!` + upgradeapi.LabelController}
 		if drain.IgnoreDaemonSets == nil || *plan.Spec.Drain.IgnoreDaemonSets {
-			args = append(args, "--delete-local-data")
+			args = append(args, "--ignore-daemonsets")
 		}
 		if drain.DeleteLocalData == nil || *drain.DeleteLocalData {
-			args = append(args, "--ignore-daemonsets")
+			args = append(args, "--delete-local-data")
 		}
```

Comment:
This is why the next milestone will see me writing the sonobuoy plugin: to catch this kind of derp early.